# Selin Ulusoy - Ödev1

In [1]:
#kütüphaneler

import pandas as pd
import numpy as np 
import seaborn as sns
import pingouin as pg

## Soru 1

### Pandas kütüphanesiyle bir DataFrame oluşturup içine Levels of Measurements'ları nominal, ordinal, interval ve ratio olan kolonlar ekleyelim.

In [2]:
#  Gender/Nominal  Mood/Ordinal  GraduationYear/Interval  Income/Ratio

dataset = { 'Gender':['F', 'M', 'M', 'F','M','F','F','M'],
            'Mood':['happy', 'very happy', 'unhappy', 'happy','very unhappy','very happy','very unhappy','happy'],
            'GraduationYear':[2016, 1998, 2010, 2018, 2019, 2003, 2013, 2015],
            'Income':['$80,000','$158,000','$97,000','$63,500','$47,600','$100,000','$58,600','$78,500']}

ds = pd.DataFrame(dataset)
ds

Gender          Mood  GraduationYear    Income
0      F         happy            2016   $80,000
1      M    very happy            1998  $158,000
2      M       unhappy            2010   $97,000
3      F         happy            2018   $63,500
4      M  very unhappy            2019   $47,600
5      F    very happy            2003  $100,000
6      F  very unhappy            2013   $58,600
7      M         happy            2015   $78,500

## Soru 2

### Ekli data üzerindeki aşağıda yer alan kolonların count, mean, std, min, 25%, 50%, 75%, max değerlerini ve IQR metoduyla outlier kalan veri sayısını çıkartıp, bu değerleri sözel ve karşılaştırmalı olarak yorumlayalım.

Öncelikle verileri bir DataFrame içinde görselleştirip yorumlayalım

In [4]:
data = pd.read_csv("data.csv")                                
df = pd.DataFrame(data) 
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

Veri seti, iki farklı tümör çeşidi olan Malignant ve Benign hücrelerinin çekirdeklerinin özelliklerini tanımlar. Malignant kanserli hücre iken Benign kanser olmayan hücredir. Bu projede  hücre çekirdeklerinin radius (merkezden çevre üzerindeki noktalara olan mesafelerin ortalaması), area (alan) ve compactness (çevre ^ 2 / alan - 1.0) özelliklerini karşılaştırmalı olarak yorumlayacağız.

#### Benign türünün count, mean, std, min, 25%, 50%, 75%, max değerlerini bulup IQR metoduyla outlier kalan veri sayısını çıkartalım

In [5]:
selected_columns = df[['radius_mean', 'area_mean', 'compactness_mean']]  
new_df= selected_columns.copy()

benign_df= new_df[df["diagnosis"]=="B"]

benign_df.describe()

radius_mean   area_mean  compactness_mean
count   357.000000  357.000000        357.000000
mean     12.146524  462.790196          0.080085
std       1.780512  134.287118          0.033750
min       6.981000  143.500000          0.019380
25%      11.080000  378.200000          0.055620
50%      12.200000  458.400000          0.075290
75%      13.370000  551.100000          0.097550
max      17.850000  992.100000          0.223900

In [7]:
columns = ['radius_mean', 'area_mean', 'compactness_mean']

Q1= benign_df[columns].quantile(0.25)
Q3= benign_df[columns].quantile(0.75)
IQR = Q3 -Q1

lower_lim = Q1 - 1.5*IQR
upper_lim = Q3 + 1.5*IQR

no_outliers_benign_df = benign_df[~((benign_df[columns]<lower_lim) | (benign_df[columns]>upper_lim)).any(axis=1)]
benign_outliers = benign_df[~(benign_df[columns].isin(no_outliers_benign_df[columns]))].reset_index(drop=True)
benign_outliers.dropna(subset=['radius_mean', 'area_mean', 'compactness_mean'],inplace=True)


print(f"Outlier değerler çıktıktan sonraki tablomuz aşağıdaki gibidir. {len(benign_outliers)} satır silindi")
no_outliers_benign_df

Outlier değerler çıktıktan sonraki tablomuz aşağıdaki gibidir. 14 satır silindi


radius_mean  area_mean  compactness_mean
19        13.540      566.3           0.08129
20        13.080      520.0           0.12700
21         9.504      273.9           0.06492
37        13.030      523.8           0.03766
46         8.196      201.9           0.05943
..           ...        ...               ...
558       14.590      657.1           0.13300
559       11.510      403.5           0.10210
560       14.050      600.4           0.11260
561       11.200      386.0           0.03558
568        7.760      181.0           0.04362

[343 rows x 3 columns]

#### Malignant türünün count, mean, std, min, 25%, 50%, 75%, max değerlerini bulup IQR metoduyla outlier kalan veri sayısını çıkartalım

In [8]:
selected_columns = df[['radius_mean', 'area_mean', 'compactness_mean']]  
new_df= selected_columns.copy()

malignant_df= new_df[df["diagnosis"]=="M"]

malignant_df.describe()

radius_mean    area_mean  compactness_mean
count   212.000000   212.000000        212.000000
mean     17.462830   978.376415          0.145188
std       3.203971   367.937978          0.053987
min      10.950000   361.600000          0.046050
25%      15.075000   705.300000          0.109600
50%      17.325000   932.000000          0.132350
75%      19.590000  1203.750000          0.172400
max      28.110000  2501.000000          0.345400

In [9]:
Q1= malignant_df[columns].quantile(0.25)
Q3= malignant_df[columns].quantile(0.75)
IQR = Q3 -Q1

lower_lim = Q1 - 1.5*IQR
upper_lim = Q3 + 1.5*IQR

no_outliers_malignant_df = malignant_df[~((malignant_df[columns]<lower_lim) | (malignant_df[columns]>upper_lim)).any(axis=1)]
malignant_outliers = malignant_df[~(malignant_df[columns].isin(no_outliers_malignant_df[columns]))].reset_index(drop=True)
malignant_outliers.dropna(subset=['radius_mean', 'area_mean', 'compactness_mean'],inplace=True)


print(f"Outlier değerler çıktıktan sonraki tablomuz aşağıdaki gibidir. {len(malignant_outliers)} satır silindi.")
no_outliers_malignant_df

Outlier değerler çıktıktan sonraki tablomuz aşağıdaki gibidir. 12 satır silindi.


radius_mean  area_mean  compactness_mean
1          20.57     1326.0           0.07864
2          19.69     1203.0           0.15990
4          20.29     1297.0           0.13280
5          12.45      477.1           0.17000
6          18.25     1040.0           0.10900
..           ...        ...               ...
562        15.22      716.9           0.20870
563        20.92     1347.0           0.22360
564        21.56     1479.0           0.11590
565        20.13     1261.0           0.10340
566        16.60      858.1           0.10230

[200 rows x 3 columns]

İstatistiksel tabloya göre Benign ve Malignant hücreleri arasındaki birtakım farklar ayırt edilebilir. min,max ve mean değerler karşılaştırıldığında, Malignant türü hücre çekirdeklerinin yarıçap (radius_mean) ve dolayısıyla alan (area_mean) değerlerinin daha yüksek bir ortalamaya sahip olduğunu gözlemlemekteyiz. Aynı durum çekirdeğin yoğunluk (compactness_mean) değerleri için de geçerlidir. Çekirdeğin radius_mean, area_mean ve compactness_mean değerlerinin, hücrenin kanser olup olmadığını anlamada ayırt edici olduğu söylenebilir.

### 3-) Ekli data üzerindeki aşağıda yer alan kolonların birbirleriyle olan korelasyonlarını, covaryanslarını, Pearson Correlation Coefficientlarını, Spearman Rank Coefficientlarını ve Effect Size'larını bularak sözel ve karşılaştırmalı olarak yorumlayalım.

#### Benign data için 

In [15]:
print("\nBenign data için korelasyon matrixi: ")
benign_df.corr()


Benign data için korelasyon matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.994435          0.060442
area_mean            0.994435   1.000000          0.052353
compactness_mean     0.060442   0.052353          1.000000

In [16]:
print("\nBenign data için kovaryans matrixi: ")
benign_df.cov()


Benign data için kovaryans matrixi: 


radius_mean     area_mean  compactness_mean
radius_mean          3.170222    237.769160          0.003632
area_mean          237.769160  18033.030100          0.237273
compactness_mean     0.003632      0.237273          0.001139

In [21]:
print("\nBenign data için spearman rank coefficient değerleri matrixi: ")
benign_df.corr(method='spearman')


Benign data için spearman rank coefficient değerleri matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.999248          0.099947
area_mean            0.999248   1.000000          0.082669
compactness_mean     0.099947   0.082669          1.000000

In [22]:
print("\nBenign data için pearson correlation coefficient değerleri matrixi: ")
benign_df.corr(method='pearson')


Benign data için pearson correlation coefficient değerleri matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.994435          0.060442
area_mean            0.994435   1.000000          0.052353
compactness_mean     0.060442   0.052353          1.000000

In [13]:
B_effectsize1= pg.compute_effsize(benign_df['area_mean'],benign_df['radius_mean'],eftype='cohen')
B_effectsize2= pg.compute_effsize(benign_df['radius_mean'],benign_df['compactness_mean'],eftype='cohen')
B_effectsize3= pg.compute_effsize(benign_df['area_mean'],benign_df['compactness_mean'],eftype='cohen')

print(f"radius_mean ile area_mean arasındaki effect size değeri = {B_effectsize1}")
print(f"radius_mean ile compactness_mean arasındaki effect size değeri = {B_effectsize2}")
print(f"compactness_mean ile area_mean arasındaki effect size değeri = {B_effectsize3}")

radius_mean ile area_mean arasındaki effect size değeri = 4.7454319505912474
radius_mean ile compactness_mean arasındaki effect size değeri = 9.582333917076165
compactness_mean ile area_mean arasındaki effect size değeri = 4.872923802607267


#### Malignant için

In [19]:
print("\nMalignant data için korelasyon matrixi: ")
malignant_df.corr()


Malignant data için korelasyon matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.990078          0.169168
area_mean            0.990078   1.000000          0.170181
compactness_mean     0.169168   0.170181          1.000000

In [20]:
print("\nMalignant data için kovaryans matrixi: ")
malignant_df.cov()


Malignant data için kovaryans matrixi: 


radius_mean      area_mean  compactness_mean
radius_mean         10.265431    1167.166503          0.029262
area_mean         1167.166503  135378.355365          3.380481
compactness_mean     0.029262       3.380481          0.002915

In [23]:
print("\nMalignant data için spearman rank coefficient değerleri matrixi: ")
malignant_df.corr(method='spearman')


Malignant data için spearman rank coefficient değerleri matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.998788          0.114265
area_mean            0.998788   1.000000          0.101147
compactness_mean     0.114265   0.101147          1.000000

In [24]:
print("\nMalignant data için pearson correlation coefficient değerleri matrixi: ")
malignant_df.corr(method='pearson')


Malignant data için pearson correlation coefficient değerleri matrixi: 


radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.990078          0.169168
area_mean            0.990078   1.000000          0.170181
compactness_mean     0.169168   0.170181          1.000000

In [25]:
M_effectsize1= pg.compute_effsize(malignant_df['area_mean'],malignant_df['radius_mean'],eftype='cohen')
M_effectsize2= pg.compute_effsize(malignant_df['radius_mean'],malignant_df['compactness_mean'],eftype='cohen')
M_effectsize3= pg.compute_effsize(malignant_df['area_mean'],malignant_df['compactness_mean'],eftype='cohen')

print(f"radius_mean ile area_mean arasındaki effect size değeri = {M_effectsize1}")
print(f"radius_mean ile compactness_mean arasındaki effect size değeri = {M_effectsize2}")
print(f"compactness_mean ile area_mean arasındaki effect size değeri = {M_effectsize3}")

radius_mean ile area_mean arasındaki effect size değeri = 3.693246109763467
radius_mean ile compactness_mean arasındaki effect size değeri = 7.6428182098023605
compactness_mean ile area_mean arasındaki effect size değeri = 3.7599485187883586


Benign ve Malignant hücre yapıları incelendiğinde, her ikisinde de radius_mean ve area_mean değerleri arasında %99 oranında kuvvetli bir korelasyon ilişkisi olduğunu gözlemlerken, compactness_mean değerinin radius_mean ve area_mean ile çok düşük korelasyon ilişkisine sahip olduğunu görüyoruz. Radius ve area arasındaki ilişki bilindiğinden sonuçlar şaşırtıcı değildir. Ancak compactness_mean korelasyon değerleri Benign hücrelerde %5 ve %6 oranında iken, Malignant hücrelerde %16 ve %17 oranına çıkmaktadır. Bu sebeple compactness_mean değerinin korelasyon ilişkilerinin, iki hücre için düşük oranda ayırt edici olduğu söylenebilir. Benzer yorum Spearman Rank Coefficient ve Pearson Correlation Coefficient değerleri için de yapılabilir.

### 4-) Bayes Teoreminin gerçek hayatta karşımıza çıkan ve kullanılan örneklerinden 5 tane bulalım.

1)  S: Aşağıdaki istatistikler göz önüne alındığında, bir kadının pozitif mamografi sonucu varsa kanser olma olasılığı nedir?
    50 Yaşın üzerindeki kadınların yüzde birinde meme kanseri var.
    Meme kanseri olan kadınların yüzde doksanı mamogramlarda pozitif test alıyor.
    Kadınların yüzde sekizi yanlış pozitif olacak.

    P(A) = 0,01
    P (~ A) = 0,99
    P (X|A) = 0.9
    P (X |~ A) = 0,08

    (0.9 * 0.01) / ((0.9 * 0.01) + (0.08 * 0.99) = 0.10.

    Pozitif test sonucu verilen bir kadının kanser olma olasılığı % 10'dur.

2) 
    Arkadaşınızın internette tanıştığı bir kızla çıkmaya başladığını hayal edin. Yeni kız arkadaşını hoş, güzel ve içine kapanık olarak tanımlıyor. Bu bilgilere dayanarak, arkadaşınızın yeni kız arkadaşının barmen, kütüphaneci veya satış elemanı olma ihtimalleri hakkında ne düşünüyorsunuz? Çoğu insan bu soruyu okuduğunda ve kızın tanımını gördüğünde, “kütüphaneci” olmasını mantıklı bir seçim gibi görüyor. Ancak, kütüphanecilerin sayısı barmenler ve satış görevlilerine kıyasla oldukça düşüktür.Bayes teoremine göre arkadaşınızın kız arkadaşının kütüphaneci olma olasılığı %0.14 olacaktır.

![](bayes1.jpg)



3) Belirli bir ilacı kullanmaya yönelik bir test, uyuşturucu kullanıcıları için %97 gerçek pozitif sonuçlar ve uyuşturucu kullanmayan kullanıcılar için %95 gerçek negatif sonuçlar üretmektedir. Genel nüfusun %0,5'inin uyuşturucu kullanıcısı olduğunu da biliyoruz. Rastgele seçilen ve testi pozitif çıkan bir kişinin uyuşturucu kullanıcısı olma olasılığı nedir? ![](bayes2.jpg)

4) A, “Hastanın karaciğer hastalığına sahip olduğu" anlamına gelsin. Geçmiş veriler size kliniğinize giren hastaların % 10'unun karaciğer hastalığı olduğunu söylüyor. P (A) = 0,10. 
    B, “Hastanın alkolik olduğu" anlamına gelsin. Kliniğe gelen hastalarının yüzde beşi alkoliktir. P(B) = 0.05. 
    Karaciğer hastalığı tanısı alan hastaların % 7'sinin alkolik olduğunu da biliyoruz. 
    P(A/B) = (0.07 * 0.1)/0.05 = 0.14 
    Hasta alkolik ise, karaciğer hastalığına yakalanma şansı % 14'tür. 
    
5) insanların % 1'inde belirli bir genetik kusur var. Gen için yapılan testlerin % 90'ı gerçek pozitif, % 9.6'sı yanlış pozitif.Bir kişi pozitif bir test sonucu alırsa, aslında genetik kusuru olma ihtimali nedir?
    A = hatalı gene sahip olma şansı. Bu soruda % 1 olarak verilmiştir. Bu aynı zamanda gene (~ A) sahip olmama olasılığının % 99 olduğu anlamına gelir.
    X = Pozitif test sonucu. P(A | X) = Genin pozitif test sonucu verme olasılığı.
    P (X | A) = Kişinin aslında gene sahip olduğu göz önüne alındığında pozitif bir test sonucu olasılığı. Bu soruda % 90 olarak verilmiştir.
    p (X / ~ A) = Kişinin geni yoksa pozitif test şansı. Bu soruda 9.6 olarak verildi
    P(A/X) = (.9 * .01) / (.9 * .01 + .096 * .99) = 0.0865 (8.65%). Hatalı genin testte bulunma olasılığı % 8.65'tir.

### 5-) Poisson, Normal , Gamma, Weilbull ve Logaritmik dağılımların benzer ve farklı yönlerini bir tablo halinde çıkaralım.(Araştırma Ödevi)

![](soru4-tablo.jpg)